File: src/types_op.c
#include "types_op.h"


void op_type(struct ProgramState *state, struct ExceptionHandler *jbuff){
    if (state->stack->next == 0)
        RAISE(jbuff, StackUnderflow);
    struct StackElem elem;
    elem.type = Type;
    elem.val.ival = state->stack->content[state->stack->next - 1].type;
    push_Stack(state->stack, elem, jbuff);
}

void op_INSTR(struct ProgramState *state, struct ExceptionHandler *jbuff){
    struct StackElem elem;
    elem.type = Type;
    elem.val.ival = Instruction;
    push_Stack(state->stack, elem, jbuff);
}

void op_INT(struct ProgramState *state, struct ExceptionHandler *jbuff){
    struct StackElem elem;
    elem.type = Type;
    elem.val.ival = Integer;
    push_Stack(state->stack, elem, jbuff);
}

void op_FLOAT(struct ProgramState *state, struct ExceptionHandler *jbuff){
    struct StackElem elem;
    elem.type = Type;
    elem.val.ival = Floating;
    push_Stack(state->stack, elem, jbuff);
}

void op_STR(struct ProgramState *state, struct ExceptionHandler *jbuff){
    struct StackElem elem;
    elem.type = Type;
    elem.val.ival = String;
    push_Stack(state->stack, elem, jbuff);
}

void op_BOOL(struct ProgramState *state, struct ExceptionHandler *jbuff){
    struct StackElem elem;
    elem.type = Type;
    elem.val.ival = Boolean;
    push_Stack(state->stack, elem, jbuff);
}

void op_TYPE(struct ProgramState *state, struct ExceptionHandler *jbuff){
    struct StackElem elem;
    elem.type = Type;
    elem.val.ival = Type;
    push_Stack(state->stack, elem, jbuff);
}

void op_NONE(struct ProgramState *state, struct ExceptionHandler *jbuff){
    struct StackElem elem;
    elem.type = Type;
    elem.val.ival = None;
    push_Stack(state->stack, elem, jbuff);
}

void op_STACK(struct ProgramState *state, struct ExceptionHandler *jbuff){
    struct StackElem elem;
    elem.type = Type;
    elem.val.ival = InnerStack;
    push_Stack(state->stack, elem, jbuff);
}

File: src/types_op.rs
//
// types_op.rs - Type operations
//

use crate::error::{Result, SscriptError};
use crate::stack::{ElemType, Stack, StackElem};

pub fn op_type(stack: &mut Stack) -> Result<()> {
    let elem = stack.peek().ok_or(SscriptError::StackUnderflow)?;
    let t = elem.elem_type();
    stack.push(StackElem::Type(t));
    Ok(())
}

pub fn op_type_instr(stack: &mut Stack) {
    stack.push(StackElem::Type(ElemType::Instruction));
}

pub fn op_type_int(stack: &mut Stack) {
    stack.push(StackElem::Type(ElemType::Integer));
}

pub fn op_type_float(stack: &mut Stack) {
    stack.push(StackElem::Type(ElemType::Floating));
}

pub fn op_type_bool(stack: &mut Stack) {
    stack.push(StackElem::Type(ElemType::Boolean));
}

pub fn op_type_str(stack: &mut Stack) {
    stack.push(StackElem::Type(ElemType::String));
}

pub fn op_type_type(stack: &mut Stack) {
    stack.push(StackElem::Type(ElemType::Type));
}

pub fn op_type_none(stack: &mut Stack) {
    stack.push(StackElem::Type(ElemType::None));
}

pub fn op_type_stack(stack: &mut Stack) {
    stack.push(StackElem::Type(ElemType::InnerStack));
}
